### PR TITLE
Compilation cache for JIT functions

### PR DESCRIPTION
**Pull request by pitrou**
This PR adds a `cache` option to the `@jit` decorator (default: false). When true, the bitcode resulting from compiling each signature is stored and looked up in a file cache. The cache is managed inside the `__pycache__` subdirectory.

Typical speedup of loading from the cache is 8x (for example down from 250ms to 30ms).

Limitation: the following functions won't be cached, a warning will be emitted instead:
- functions calling a ctypes/cffi function pointer
- functions with lifted loops

### CODE REVIEW

File: numba/pythonapi.py [hunk 1]
```diff
@@ -1246,10 +1242,11 @@ def create_np_timedelta(self, val, unit_code):
         fn = self._get_function(fnty, name="numba_create_np_timedelta")
         return self.builder.call(fn, [val, unit_code])
 
-    def recreate_record(self, pdata, size, dtypeaddr):
+    def recreate_record(self, pdata, size, dtype):
         fnty = Type.function(self.pyobj, [Type.pointer(Type.int(8)),
                                           Type.int(), self.pyobj])
         fn = self._get_function(fnty, name="numba_recreate_record")
+        dtypeaddr = self.unserialize(self.serialize_object(dtype))
```
Comment (pitrou) [inline numba/pythonapi.py]: This is unfortunately slow (e.g. 6µs for a simple two-member struct type).

**Comment by pitrou** [comment]
For those looking to review, the meat of the caching logic is here:
https://github.com/numba/numba/pull/1051/files#diff-067c6329c9ddf60932fd3f3111969487R62
https://github.com/numba/numba/pull/1051/files#diff-66adccf97cb99c68d3c33d8025531cc6R437

**Comment by seibert** [comment]
I think we need some way to include the Numba version (or some counter we can increment between releases) that generated the code in the cache key.  It is quite likely that we will need to expire items from the cache between Numba releases due to internal ABI changes, or bug fixes in the compiler.

**Comment by seibert** [comment]
Also, we should document how to clear the function cache.

**Comment by pitrou** [comment]
> I think we need some way to include the Numba version (or some counter we can increment between releases) that generated the code in the cache key

The Numba version is already included in the index file header.

> Also, we should document how to clear the function cache.

That would be "remove the ``__pycache__`` directory". I'm not sure we want anything more sophisticated...

**Comment by pitrou** [comment]
Or do you mean add a clear_cache() method?

**Comment by seibert** [comment]
I don't know if we need a programmatic way to clear the cache yet, though we do want to ensure our unit tests don't accidentally use cached versions of the functions.

**Comment by pitrou** [comment]
Oh, well, they won't, since caching is opt-in :)

**Comment by pitrou** [comment]
Wait, I wrote `cache=True` in the docs... must fix that.

**Comment by sklam** [comment]
The following testcase will load the wrong cache due to name collision for the nested function `foo`:
https://gist.github.com/sklam/48fd3e698afcac93628e

**Comment by sklam** [comment]
A simpler version of the problem: https://gist.github.com/sklam/b694cc1231614ec5a67c

**Comment by pitrou** [comment]
Hmm, I'll take a look. It should be reasonable to disambiguate by line number.

**Comment by pitrou** [comment]
Should be fixed. Note this may not be sufficient for closures, though.

**Comment by seibert** [comment]
I'm worried about the closure case since we have shown off function factories in the past.  Is there a way to detect if we are compiling in a closure and disable caching even if the user has selected it?

**Comment by pitrou** [comment]
One way to do that would be to only compile global functions.

**Comment by seibert** [comment]
I think that is a reasonable restriction for now, and we can print a warning if the user tries to cache a non-global function.  We can always loosen the restriction in the future if we find a safe way to handle the situation.

**Comment by pitrou** [comment]
Ok, I've added a detection so that closures don't get cached.

**Pull request merged**